repos/system_upgrade/common/actors/initramfs/checkfipsenabled/actor.py

```diff
@@ -0,0 +1,32 @@
+from leapp.actors import Actor
+from leapp.exceptions import StopActorExecutionError
+from leapp.libraries.common import mounting
+from leapp.models import FIPSInfo, TargetUserSpaceInfo
+from leapp.tags import IPUWorkflowTag, LateTestsPhaseTag
+
+
+class CheckFIPSCorrectlyEnabled(Actor):
+    """
+    Stop the upgrade if FIPS was not correctly enabled after rebooting into the upgrade initramfs.
+    """
+
+    name = 'check_fips_correctly_enabled'
+    consumes = (FIPSInfo,)
+    produces = ()
+    tags = (LateTestsPhaseTag, IPUWorkflowTag)
+
+    def process(self):
+        fips_info = next(self.consume(FIPSInfo), None)
+        if not fips_info:
+            # Unexpected, FIPSInfo is produced unconditionally
+            raise StopActorExecutionError('Cannot check for the correct FIPS state in the upgrade initramfs',
+                                          details={'Problem': 'Did not receive any FIPSInfo message'})
+
+        if fips_info.is_enabled:
+            # @Todo: checking for fips=1 in cmdline is likely insufficient to determine whether the upgrade environment
```

Comment:
I don't think there's a simple way to run the check we would need to run, since AFAIK all of these tests run *inside* of the `systemd-nspawn` container, right? That did already contain everything that was needed, and FIPS mode does already work correctly in this environment.

In this case, I think verifying that FIPS mode was enabled is good enough – it'll at least fail when fips=1 was not correctly passed into the kernel command line, although we could probably get the same assurance by adding a unit test for the `addupgradebootentry` actor that verifies that `fips=1` will be preserved.

In any case, if we keep this test, it should read `/proc/sys/crypo/fips_enabled` and check that it contains `1`, since that will tell us whether the kernel (a) supports FIPS mode, and (b) has successfully been switched into FIPS mode.

Reply:
You are correct that leapp runs inside the container at the time these tests are executed. Historically, we have encountered issues when using grubby to manipulate boot entries due to which I am being rather cautious as booting without `fips=1` would result in the target system without `fips=1` as well (not sure how bad the consequences are).

Reply:
If the system boots without `fips=1`, that means FIPS mode was essentially disabled silently, so we should really make sure that doesn't happen.

However, `fips=1` on the kernel command line and `1` in `/proc/sys/crypto/fips_enabled` have the same meaning, and the proc file is simpler to parse, so I'd suggest to keep this test but check the contents of `/proc/sys/crypto/fips_enabled`.

Reply:
Thank you for your suggestion to use `/proc/sys/crypto/fips_enabled` instead of parsing cmdline, I've changed the code accordingly.